halo2_proofs/src/plonk/vanishing/prover.rs

```diff
@@ -42,10 +49,43 @@ impl<C: CurveAffine> Argument<C> {
         transcript: &mut T,
     ) -> Result<Committed<C>, Error> {
         // Sample a random polynomial of degree n - 1
-        let mut random_poly = domain.empty_coeff();
-        for coeff in random_poly.iter_mut() {
-            *coeff = C::Scalar::random(&mut rng);
-        }
+        #[cfg(feature = "multicore")]
+        let random_poly = {
+            let n_threads = current_num_threads();
+            let n = 1usize << domain.k as usize;
+            let n_chunks = n_threads + if n % n_threads != 0 { 1 } else { 0 };
+            let mut rand_vec = vec![C::Scalar::ZERO; n];
+
+            let mut thread_seeds: Vec<ChaCha20Rng> = (0..n_chunks)
+                .into_iter()
+                .map(|_| {
+                    let mut seed = [0u8; 32];
+                    rng.fill_bytes(&mut seed);
+                    ChaCha20Rng::from_seed(seed)
```

Comment:
I believe this could safely use `ChaCha12Rng` while conservatively meeting our security requirements. The paper [Too Much Crypto](https://eprint.iacr.org/2019/1492) argues in favour of ChaCha8; I do not entirely endorse all of the conclusions of that paper, but there is no significant doubt in my mind that 12 rounds is sufficient. There is a more recent (Nov 2021) paper on cryptanalysis of ChaCha, [PNB-focused Differential Cryptanalysis of ChaCha Stream Cipher](https://eprint.iacr.org/2021/1537) than the ones considered in Too Much Crypto (which is originally from Dec 2019), but it does not improve on brute force and in any case would only extend to 7.25 rounds if it did.